src/client/mod.rs

```diff
@@ -32,39 +32,120 @@ use status::StatusClass::Redirection;
 use {Url, HttpResult};
 use HttpError::HttpUriError;
 
+use self::response::{HttpResponse};
+use self::request::{FreshHttpRequest, StreamingHttpRequest};
+use net::{Fresh};
+
 pub use self::request::Request;
 pub use self::response::Response;
 
 pub mod request;
 pub mod response;
 
+/// The struct represents parts of an HTTP request that are prepared for
+/// sending to the server.
+pub struct RequestTemplate<'a> {
+    pub url: Url,
+    pub method: Method,
+    pub body: Option<Body<'a>>,
+    pub headers: Option<Headers>,
+    pub can_have_body: bool,
+}
+
+/// A trait that represents the functionality of creating a brand new fresh
+/// request (an instance of a type implementing `FreshHttpRequest`).
+///
+/// Each particular implementation of the trait can provide a different way
+/// for achieving this, as well as a custom type of request that it produces.
+///
+/// Therefore, by defining the type of the fresh request that it will produce,
+/// it also directly defines the type of the response, since the chain of
+/// `Fresh -> Streaming -> Response` types is always well defined given a
+/// particular `FreshHttpRequest` starting point.
+///
+/// TODO: Might need a better name?
+pub trait HttpRequestFactory {
+    /// The type of the `FreshHttpRequest` that this factory will produce.
+    type RequestType: FreshHttpRequest;
+    /// Creates a new instance of the particular `HttpRequestFactory`.
+    /// This method is a convenience so that the `Client` can still be
+    /// instantiated without providing a factory instance, but with a simple
+    /// `new()` call. This, however, means that all particular
+    /// `HttpRequestFactory` implementations must have a set of default
+    /// parameters with which they can be created.
+    fn new() -> Self;
```

Comment:
We could get rid of this and just have `HttpRequestFactory` inherit from `Default`.